File: labs/J70_State/python/complete/orderstatus.py



class OrderStatus():
    def to_string(self) -> str:
        pass
    def next(self, nextStatus: object) -> object:
        pass


class OrderStatuses():
    created:OrderStatus
    filled:OrderStatus
    shipped:OrderStatus
    billed:OrderStatus



class CreatedStatus(OrderStatus):
    def to_string(self) -> str:
        return "Created"
    def next(self,nextStatus)->OrderStatus:
        if nextStatus != OrderStatuses.filled:
            raise Exception('Bogus State Transition!')
        return nextStatus

OrderStatuses.created = CreatedStatus()




class FilledStatus(OrderStatus):
    def to_string(self) -> str:
        return "Filled"
    def next(self,nextStatus)->OrderStatus:
        if nextStatus != OrderStatuses.shipped:
            raise Exception('Bogus State Transition!')
        return nextStatus

OrderStatuses.filled=FilledStatus()






class ShippedStatus(OrderStatus):
    def to_string(self) -> str:
        return "Shipped"
    def next(self,nextStatus)->OrderStatus:
        if nextStatus != OrderStatuses.billed:
            raise Exception('Bogus State Transition!')
        return nextStatus

OrderStatuses.shipped=ShippedStatus()




class BilledStatus(OrderStatus):
    def to_string(self) -> str:
        return "Billed"
    def next(self,nextStatus)->OrderStatus:
        raise Exception('Bogus State Transition!')

OrderStatuses.billed=BilledStatus()

File: labs/J70_State/java/complete/app/src/main/java/StateComplete/OrderStatus.java
package StateComplete;

public abstract class OrderStatus {
  public abstract OrderStatus next(OrderStatus newStatus) throws Exception ;

  public static final OrderStatus created = new Created();
  public static final OrderStatus shipped = new Shipped();
  public static final OrderStatus filled = new Filled();
  public static final OrderStatus billed = new Billed();
}

class Created extends OrderStatus{
  
  @Override
  public String toString(){
    return "Created";
  }

  @Override
  public OrderStatus next(OrderStatus newStatus) throws Exception{
    if(newStatus != OrderStatus.filled){
      throw new Exception("Bogus transition");
    }
    return newStatus;
  }
}

class Filled extends OrderStatus{
  
  @Override
  public String toString(){
    return "Filled";
  }

  @Override
  public OrderStatus next(OrderStatus newStatus) throws Exception {
    if(newStatus != OrderStatus.shipped){
      throw new Exception("Bogus transition");
    }
    return newStatus;
  }
}


class Shipped extends OrderStatus{
  
  @Override
  public String toString(){
    return "Shipped";
  }

  @Override
  public OrderStatus next(OrderStatus newStatus) throws Exception {
    if(newStatus != OrderStatus.billed){
      throw new Exception("Bogus transition");
    }
    return newStatus;
  }
}

class Billed extends OrderStatus{
  
  @Override
  public String toString(){
    return "Billed";
  }

  @Override
  public OrderStatus next(OrderStatus newStatus) throws Exception {
    throw new Exception("Bogus transition");
  }
}